### PR TITLE
Update coupon API date handling

### PR DESCRIPTION
**Pull request by claudiulodro**
Relates to #13525.

This PR is for the branch I'm using in #13587. It should be easier to review the API stuff on its own, instead of everything as one giant wall of new code.

I've updated all of the coupon APIs to work nicely with the new date implementation.

### CODE REVIEW

File: includes/api/class-wc-rest-coupons-controller.php [hunk 1]
```diff
@@ -330,6 +377,11 @@ public function get_item_schema() {
 					'type'        => 'string',
 					'context'     => array( 'view', 'edit' ),
 				),
+				'date_expires_gmt' => array(
```
Comment (claudiosanches) [inline includes/api/class-wc-rest-coupons-controller.php]: Now have 2 `UTC` dates here.
Better change to:

`date_expire` = `The date the coupon expire, in the site's timezone.`
`date_expire_gmt` = `The date the coupon expire, as GMT.`

File: includes/api/v1/class-wc-rest-coupons-controller.php [hunk 2]
```diff
@@ -416,13 +419,13 @@ public function get_item_schema() {
...
 				'date_created' => array(
-					'description' => __( "The date the coupon was created, in the site's timezone.", 'woocommerce' ),
+					'description' => __( "The date the coupon was created, as GMT.", 'woocommerce' ),
 					'type'        => 'date-time',
 					'context'     => array( 'view', 'edit' ),
 					'readonly'    => true,
 				),
 				'date_modified' => array(
-					'description' => __( "The date the coupon was last modified, in the site's timezone.", 'woocommerce' ),
+					'description' => __( "The date the coupon was last modified, as GMT.", 'woocommerce' ),
```
Comment (claudiosanches) [inline includes/api/v1/class-wc-rest-coupons-controller.php]: On v1 needs to keep as `site's timezone` unless we already returned it as GMT before.

File: includes/api/v1/class-wc-rest-coupons-controller.php [hunk 3]
```diff
@@ -446,7 +449,7 @@ public function get_item_schema() {
 				),
 				'expiry_date' => array(
 					'description' => __( 'UTC DateTime when the coupon expires.', 'woocommerce' ),
-					'type'        => 'string',
+					'type'        => 'date-time',
```
Comment (claudiosanches) [inline includes/api/v1/class-wc-rest-coupons-controller.php]: Need to keep the old format here, only update on v2.

**Review by claudiosanches — CHANGES_REQUESTED** [review]

File: includes/api/class-wc-rest-coupons-controller.php (hunk 1 above)
Comment (mikejolley) [inline includes/api/class-wc-rest-coupons-controller.php]: ^ `date_expires_gmt ` and `date_expires_gmt`

File: includes/api/v1/class-wc-rest-coupons-controller.php (hunk 2 above)
Comment (mikejolley) [inline includes/api/v1/class-wc-rest-coupons-controller.php]: +1 keep the v1 api unchanged.

File: includes/api/class-wc-rest-coupons-controller.php [hunk 4]
```diff
@@ -149,14 +141,57 @@ public function prepare_object_for_response( $object, $request ) {
...
+			'limit_usage_to_x_items' => $data['limit_usage_to_x_items'],
+			'free_shipping' => $data['free_shipping'],
+			'product_categories' => $data['product_categories'],
+			'excluded_product_categories' => $data['excluded_product_categories'],
+			'exclude_sale_items' => $data['exclude_sale_items'],
+			'minimum_amount' => $data['minimum_amount'],
+			'maximum_amount' => $data['maximum_amount'],
+			'email_restrictions' => $data['email_restrictions'],
+			'used_by' => $data['used_by'],
+			'meta_data' => $data['meta_data'],
```
Comment (mikejolley) [inline includes/api/class-wc-rest-coupons-controller.php]: Pedantic comment,  but I like using https://atom.io/packages/atom-alignment :) Maybe there is one for sublime? Or you can try Atom :D :trollface:

File: includes/api/class-wc-rest-coupons-controller.php [hunk 5]
```diff
@@ -127,15 +127,7 @@ protected function get_object( $id ) {
...
-	/**
-	 * Prepare a single coupon output for response.
-	 *
-	 * @since  2.7.0
-	 * @param  WC_Data         $object  Object data.
-	 * @param  WP_REST_Request $request Request object.
-	 * @return WP_REST_Response
-	 */
-	public function prepare_object_for_response( $object, $request ) {
+	protected function get_formatted_item_data( $object ) {
```
Comment (justinshreve) [inline includes/api/class-wc-rest-coupons-controller.php]: This function still needs its own docblock.

File: includes/api/class-wc-rest-coupons-controller.php (hunk 4 above)
Comment (claudiosanches) [inline includes/api/class-wc-rest-coupons-controller.php]: On Sublime you can use https://packagecontrol.io/packages/Alignment

**Review by emdashcodes — COMMENTED** [review]
This is the only issue I'm seeing. PHPUnit tests all pass/look good.

File: includes/api/class-wc-rest-coupons-controller.php (hunk 4 above)
Comment (claudiosanches) [inline includes/api/class-wc-rest-coupons-controller.php]: Instructions here: https://www.granneman.com/webdev/editors/sublime-text/packages/how-to-install-and-use-sublime-alignment/

**Pull request closed without merging**